File: 150. Evaluate Reverse Polish Notation/evalRPN.py
class Solution:
    operators = ["+", "-", "*", "/"]
    def evalRPN(self, tokens) -> int:
        stack = []

        for ch in tokens:
            if ch in self.operators:
                secondOperand = stack.pop()
                firstOperand = stack.pop()

                result = self.evalExpression(firstOperand, secondOperand, ch)
                stack.append(result)
            else:
                stack.append(int(ch))
        
        return stack.pop()


    def evalExpression(self, first, second, operator):
        if operator == '+':
            return first + second
        elif operator == '-':
            return first - second
        elif operator == '*':
            return first * second
        else:
            return int(first/second)


sol = Solution()

print(sol.evalRPN(["10","6","9","3","+","-11","*","/","*","17","+","5","+"]))
        

File: 150. Evaluate Reverse Polish Notation/eval_rpn.rs
impl Solution {
    
    // O(n) time,
    // O(n) space,
    // Approach: stack, math
    pub fn eval_rpn(tokens: Vec<String>) -> i32 {
        
        let mut stack: Vec<i32> = Vec::new();
        
        for token in tokens {
            
            if token.parse::<i32>().is_ok() {
                stack.push(token.parse::<i32>().unwrap());
            }
            else {
                let val1 = stack.pop().unwrap();
                let val2 = stack.pop().unwrap();
                
                if token == "+" {
                    stack.push(val1+val2);
                }
                else if token == "-" {
                    stack.push(val2-val1);
                }
                else if token == "*" {
                    stack.push(val1*val2);
                }
                else {
                    stack.push(val2/val1);
                }
            }
            
        }
        
        return stack[0];
    }
}